bin/bats

```diff
@@ -2,68 +2,53 @@
 
 set -euo pipefail
 
-BATS_READLINK='true'
-BATS_REALPATH=''
-if command -v 'greadlink' >/dev/null; then
-  BATS_READLINK='greadlink'
-elif command -v 'readlink' >/dev/null; then
-  BATS_READLINK='readlink'
+if command -v greadlink >/dev/null; then
+  bats_readlinkf() {
+    greadlink -f "$1"
+  }
+else
+  bats_readlinkf() {
+    readlink -f "$1"
+  }
 fi
 
-if ! "${BATS_READLINK}" -m "$0" &>/dev/null; then
-  BATS_REALPATH='realpath'
-fi
-
-resolve_path() {
-  local path_to_bats=$1
-  local target_path
-
-  if [[ -L "${path_to_bats}" ]]; then
-    # Busybox has no readlink from coreutils
-    if [[ -n "$BATS_REALPATH" ]]; then
-      target_path="$("$BATS_REALPATH" "$path_to_bats")"
-      # we need to traverse the paths
-      target_path="$(dirname "$target_path")" # strip bats
-      target_path="$(dirname "$target_path")" # strip bin
-    else
-      target_path="$("$BATS_READLINK" -m "$path_to_bats/../../")"
-    fi
-  elif [[ -f "${path_to_bats}" ]]; then
-    target_path=$(dirname "$path_to_bats")
-    if [[ -n "$BATS_REALPATH" ]]; then
-      target_path="$("$BATS_REALPATH" "$target_path")"
-    else
-      target_path="$("$BATS_READLINK" -m "$target_path")"
-    fi
-    target_path="$(dirname "$target_path")" # strip bin
-  elif [[ -d "${path_to_bats}" ]]; then
-    if [[ -n "$BATS_REALPATH" ]]; then
-      target_path="$("$BATS_REALPATH" "$path_to_bats")"
-      target_path="$(dirname "$target_path")"
-    else
-      target_path="$("$BATS_READLINK" -m "$path_to_bats/../")"
-    fi
-  fi
-
-  echo "$target_path"
+fallback_to_readlinkf_posix() {
+  bats_readlinkf() {
+    [ "${1:-}" ] || return 1
+    target=$1 max_symlinks=10 CDPATH=''
+
+    [ -e "${target%/}" ] || target=${1%"${1##*[!/]}"} # trim trailing slashes
+    [ -d "${target:-/}" ] && target="$target/"
+
+    cd -P . 2>/dev/null || return 1
+    while [ "$max_symlinks" -gt 0 ] && max_symlinks=$((max_symlinks - 1)); do
+      if [ ! "$target" = "${target%/*}" ]; then
+        cd -P "${target%/*}/" 2>/dev/null || break
+        target=${target##*/}
```

Comment:
Do these direct-to-variable assignments need quoting? We tend to quote by default out of paranoia!